plugin/rewrite/reverter.go

```diff
@@ -21,22 +22,77 @@ type ResponseRule struct {
 // dig will complain with ';; Question section mismatch: got example.org/HINFO/IN'
 type ResponseReverter struct {
 	dns.ResponseWriter
+	request          *dns.Msg
 	originalQuestion dns.Question
+	originalOPT      *dns.OPT
 	ResponseRewrite  bool
 	ResponseRules    []ResponseRule
 }
 
+// return a deep copy of the OPT record
+func dupOPT(opt *dns.OPT) (o *dns.OPT) {
+	edns0 := make([]dns.EDNS0, len(opt.Option))
+	for i, e := range opt.Option {
+		edns0[i] = e.Copy()
+	}
+	o = &dns.OPT{opt.Hdr, edns0}
+	return o
+}
+
+// return a deep copy of the OPT inside this msg
+func copyOPT(m *dns.Msg) (o *dns.OPT) {
+	for i := len(m.Extra) - 1; i >= 0; i-- {
+		if m.Extra[i].Header().Rrtype == dns.TypeOPT {
+			if opt, ok := m.Extra[i].(*dns.OPT); ok {
+				return dupOPT(opt)
+			}
+		}
+	}
+	return o
+}
+
+// replace the OPT inside the msg with the a copy of the one proposed
+// manage properly if OPT record is nil
+func replaceOPT(m *dns.Msg, o *dns.OPT) {
+	for i := len(m.Extra) - 1; i >= 0; i-- {
+		if m.Extra[i].Header().Rrtype == dns.TypeOPT {
+			if o == nil {
+				copy(m.Extra[i:], m.Extra[i+1:])
+				m.Extra[len(m.Extra)-1] = nil
+				m.Extra = m.Extra[:len(m.Extra)-1]
+				break
+			}
+			m.Extra[i] = dupOPT(o)
+			break
+		}
+	}
+	if o != nil {
+		m.Extra = append(m.Extra, dupOPT(o))
+	}
+}
+
 // NewResponseReverter returns a pointer to a new ResponseReverter.
 func NewResponseReverter(w dns.ResponseWriter, r *dns.Msg) *ResponseReverter {
 	return &ResponseReverter{
 		ResponseWriter:   w,
+		request:          r,
 		originalQuestion: r.Question[0],
+		originalOPT:      copyOPT(r),
 	}
 }
 
 // WriteMsg records the status code and calls the underlying ResponseWriter's WriteMsg method.
 func (r *ResponseReverter) WriteMsg(res *dns.Msg) error {
 	res.Question[0] = r.originalQuestion
+
+	// set back the client request to its original OPT record (before changed by Rewrite)
+	replaceOPT(r.request, r.originalOPT)
```

Comment:
not sure if it make sense to restore original OPT in request. It's only important to remove the OPT record if it was not originally present in request, so that `SizeAndDo()` to not insert OPT record into response

Reply:
as if today it is important : you get the size of the buffer in the OPT header.
(used in SizeAndDo).